File: src/cquill_ffi.erl
-module(cquill_ffi).
-export([coerce_value/1, escape_sql_string/1, safe_tuple_size/1, tuple_to_list/1]).

%% Coerce a Gleam value to a tagged union for type detection
%% This allows the query builder to create the appropriate Value type
coerce_value(Value) when is_integer(Value) ->
    {int_val, Value};
coerce_value(Value) when is_float(Value) ->
    {float_val, Value};
coerce_value(Value) when is_binary(Value) ->
    {string_val, Value};
coerce_value(true) ->
    {bool_val, true};
coerce_value(false) ->
    {bool_val, false};
coerce_value(nil) ->
    nil_val;
coerce_value(none) ->
    nil_val;
coerce_value(Value) when is_list(Value) ->
    {list_val, [coerce_value(V) || V <- Value]};
coerce_value(_) ->
    unknown_val.

%% Escape single quotes in SQL strings by doubling them
escape_sql_string(String) when is_binary(String) ->
    binary:replace(String, <<"'">>, <<"''">>, [global]).

%% Safely get the size of a tuple, returns -1 if not a tuple
safe_tuple_size(Value) when is_tuple(Value) ->
    erlang:tuple_size(Value);
safe_tuple_size(_) ->
    -1.

%% Convert a value to a list suitable for row representation
%% - Tuples are converted using erlang:tuple_to_list/1
%% - Lists are returned as-is (in case pog already converted)
%% - Other values are wrapped in a single-element list
tuple_to_list(Value) when is_tuple(Value) ->
    erlang:tuple_to_list(Value);
tuple_to_list(Value) when is_list(Value) ->
    Value;
tuple_to_list(Value) ->
    [Value].

File: src/cquill_ffi.mjs
// JavaScript FFI for cquill query builder
// Provides value coercion for the query DSL

export function coerce_value(value) {
  if (typeof value === 'number') {
    if (Number.isInteger(value)) {
      return { type: 'IntVal', 0: value };
    } else {
      return { type: 'FloatVal', 0: value };
    }
  }

  if (typeof value === 'string') {
    return { type: 'StringVal', 0: value };
  }

  if (typeof value === 'boolean') {
    return { type: 'BoolVal', 0: value };
  }

  if (value === null || value === undefined) {
    return { type: 'NilVal' };
  }

  if (Array.isArray(value)) {
    return { type: 'ListVal', 0: value.map(coerce_value) };
  }

  // Handle Gleam's None type
  if (value && value.type === 'None') {
    return { type: 'NilVal' };
  }

  return { type: 'UnknownVal' };
}

// Escape single quotes in SQL strings by doubling them
export function escape_sql_string(str) {
  return str.replace(/'/g, "''");
}

// Convert float to string
export function float_to_string(f) {
  return String(f);
}
